client/question/services/ParentPageService.js

```diff
@@ -0,0 +1,61 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service that handles all TIE interactions with its parent
+ * page.
+ */
+tie.factory('ParentPageService', [
+  '$window', 'PARENT_PAGE_URL_ORIGIN',
+  function($window, PARENT_PAGE_URL_ORIGIN) {
+    /**
+     * Used to define the raw code message type.
+     * @type {String}
+     * @constant
+     */
+    var MESSAGE_TYPE_RAW_CODE = 'raw_code';
+
+    /**
+     * Sends the message to the parent page, if it exists, via postMessage.
+     *
+     * @param {String} messageType Type of message that is being sent.
+     * @param {String} messageData Data to be sent to parent page.
+     */
+    var sendMessage = function(messageType, messageData) {
+      if (PARENT_PAGE_URL_ORIGIN) {
+        $window.parent.postMessage(
+          JSON.stringify(messageData), PARENT_PAGE_URL_ORIGIN);
+      }
+    };
+
+    return {
+      /**
+       * Sends the raw user code to the parent page, if it exists.
+       *
+       * @param {String} rawCode User code to send to parent page.
+       */
+      sendRawCode: function(rawCode) {
+        sendMessage(MESSAGE_TYPE_RAW_CODE, rawCode);
+      },
+      /**
+       * Returns the URL origin of the parent page, if it exists. Otherwise,
+       * returns null.
+       * @return {String|null}
+       */
+      getParentPageUrlOrigin: function() {
```

Comment:
Just to check -- what external method do you envision calling this? I don't think I see the need to expose it, but I might be missing something.

Reply:
Later on I'll add the button to submit user code to the parent page, which calls this method when the button is clicked.

Reply:
Ah -- I think that may not be quite right, then.

Firstly, submitToParentPage should call sendRawCode, which it does. So that doesn't interact with this.

Secondly, there is no need for the learner view directive to know what the parent page origin is. All it needs to know is whether it's currently embedded or not (and if the embedding is "valid"). This requires three things: there actually is a parent page at the moment, the url origin we specify is not null, and the parent page matches the url origin we specify. (Note that even if a url origin is specified the page may not be embedded in the current session).

So, I would replace this with something like isIframed(). It's also fine to delete it altogether from this PR, and only add it in the next.

Reply:
That makes sense. I'll remove it for now and be sure to add it to the next PR as an isIframed method instead.